compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "symbol.h"
+#include "type.h"
+#include "LoopStmt.h"
+#include "CForLoop.h"
+#include "WhileStmt.h"
+
+//helper datastructures/types
+
+//maybe more memoization than caching
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+typedef std::map<SymExpr*, std::pair<Expr*,Type*> > ActualUseDefCastMap;
+
+//prototypes
+bool isExprSafeForReorder(Expr * e);
+bool canPrimMoveCreateCommunication(CallExpr* ce);
+inline bool possibleDepInBetween(Expr* e1, Expr* e2);
+inline bool requiresCast(Type* t);
+inline bool isIntegerPromotionPrimitive(PrimitiveTag tag);
+inline bool isNonEssentialPrimitive(CallExpr* ce);
+bool isSafePrimitive(CallExpr* ce);
+
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo);
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap);
+
+void denormalize(void);
+void denormalize(FnSymbol *fn);
+void denormalize(Expr* def, SymExpr* use, Type* castTo);
+
+
+void denormalize(void) {
+  if(fDenormalize) {
+    forv_Vec(FnSymbol, fn, gFnSymbols) {
+      denormalize(fn);
+    }
+  }
+}
+
+void denormalize(FnSymbol *fn) {
```

Comment:
I'd appreciate a general comment about what denormalization is, what kind of things get denormalized, and why denormalization is useful. You have a lot of that in your PR message but I'd put it in a comment here. It can be very helpful to see a written description of the transformation being performed.